rfc/src/rfcs/0009-function-contracts.md

```diff
@@ -0,0 +1,674 @@
+   }
+   ```
+  
+   `requires` here indicates this function expects its `divisor` input to never
+   be 0, or it will not execute correctly (i.e. panic).
+
+   `ensures` puts a bound on the output, relative to the `dividend` input.
+
+   Conditions in contracts are plain Rust expressions which can reference the
+   function arguments and, in case of `ensures`, the result of the function as a
+   special `result` variable (see [open questions](#open-questions) on a
+   discussion about (re)naming). Syntactically Kani supports any Rust
+   expression, including function calls, defining types etc. However they must
+   be side-effect free[^side-effects] or Kani will throw a compile error.
+
+   Multiple `requires` and `ensures` clauses are allowed on the same function,
+   they are implicitly logically conjoined.
+
+
+2. Next, Kani must make sure that the approximation we specified actually holds
+   by **checking** the contract against the implementation. This is in contrast
+   to ["stubbing"][stubbing], where the approximation is blindly trusted.
+
+   The contract must always overapproximate the function to guarantee soundness.
+   To facilitate the check Kani needs a suitable environment to verify the
+   function in. For this proposal the environment must be provided by us (the
+   users). See [future possibilities](#future-possibilities) for a discussion
+   about the arising soundness issues and their remedies.
+
+   We provide the checking environment for our contract with a special new
+   `proof_for_contract` harness.
+
+   ```rs
+   #[kani::proof_for_contract(my_div)]
+   fn my_div_harness() {
+     my_div(kani::any(), kani::any())
+   }
+   ```
+
+   Similar to a verification harness for any other function, we are supposed to
+   create all possible input combinations the function can encounter, then call
+   the function at least once with those abstract inputs. If we forget to call
+   `my_div` Kani reports an error.
+   
+   Unlike a unit-test we can however elide any checks of the output and
+   post-call state. Instead Kani uses the conditions we specified in the
+   contract as checks. Preconditions (`requires`) are inserted as `kani::assume`
+   *before* the call to `my_div`, limiting the generic inputs to those the
+   function is actually defined for. Postconditions (`ensures`) are inserted as
+   `kani::assert` checks *after* the call to `my_div`.
+
+   The expanded version of our harness and function is equivalent to the following:
+
+   ```rs
+   #[kani::proof]
+   fn my_div_harness() {
+     let dividend = kani::any();
+     let divisor = kani::any();
+     kani::assume(divisor != 0); // requires
+     let result = my_div(dividend, divisor);
+     kani::assert(result <= dividend); // ensures
+   }
+   ```
+
+   This expanded harness is then verified like any other harness but also gives
+   the green light for the next step: verified stubbing.
+
+3. In the last phase the **verified** contract is ready for us to use to
```

Comment:
There is no real sequential dependency between these steps

Reply:
Technically true, but I think we should introduce it to prevent unsoundness. At least by default. I've added a discussion to open questions about how to circumvent this order if desired but by default I think it should be enforced. Basically encourage users to get things right.

Reply:
As the default workflow yes. For a first version you can avoid the enforcement by passing `--harness` and in open questions we discuss more ways to control contract checking which we can add later.
Thus users won't be forced to always go bottom-up, but they have to be explicit about that.